internal/provider/resource_access_model.go

```diff
@@ -0,0 +1,269 @@
+		Description: "A resource that represent a Juju Access Model.",
+
+		CreateContext: resourceAccessModelCreate,
+		ReadContext:   resourceAccessModelRead,
+		UpdateContext: resourceAccessModelUpdate,
+		DeleteContext: resourceAccessModelDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceAccessModelImporter,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"model": {
+				Description: "The name of the model for access management",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"users": {
+				Description: "List of users to grant access to",
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"access": {
+				Description: "Type of access to the model",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+		},
+	}
+}
+
+func resourceAccessModelCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*juju.Client)
+
+	var diags diag.Diagnostics
+
+	model := d.Get("model").(string)
+	access := d.Get("access").(string)
+	usersInterface := d.Get("users").([]interface{})
+	users := make([]string, len(usersInterface))
+	for i, v := range usersInterface {
+		users[i] = v.(string)
+	}
+
+	uuid, err := client.Models.ResolveModelUUID(model)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	modelUUIDs := []string{uuid}
+
+	for _, user := range users {
+		err := client.Models.GrantModel(juju.GrantModelInput{
+			User:       user,
+			Access:     access,
+			ModelUUIDs: modelUUIDs,
+		})
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", model, access))
+
+	return diags
+}
+
+func resourceAccessModelRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*juju.Client)
+
+	var diags diag.Diagnostics
+
+	id := strings.Split(d.Id(), ":")
+	usersInterface := d.Get("users").([]interface{})
+	stateUsers := make([]string, len(usersInterface))
+	for i, v := range usersInterface {
+		stateUsers[i] = v.(string)
+	}
+
+	uuid, err := client.Models.ResolveModelUUID(id[0])
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	response, err := client.Users.ModelUserInfo(uuid)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set("model", id[0]); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("access", id[1]); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var users []string
+
+	for _, user := range stateUsers {
+		for _, modelUser := range response.ModelUserInfo {
+			if user == modelUser.UserName && string(modelUser.Access) == id[1] {
+				users = append(users, modelUser.UserName)
+			}
+		}
+	}
+
+	if err = d.Set("users", users); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+// Updating the access model supports three cases
```

Comment:
Docs for functions in go must start with the function name `resourceAccessModelUpdate` in this case.

Reply:
That was more of an internal note than external facing document, though I can change the paragraph and make it appropriate.

Reply:
OK in that case, the comment should not be adjacent to the function so it is not captured by the godoc.